### PR TITLE
PLG-97: Track product changes

### PR DESCRIPTION
**Pull request by nedick007**
Tracking functionality for product change from system administration.

### CODE REVIEW

**Comment by nedick007** [comment]
@adamof ready for review

File: Observer/Product.php [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+                                       'metadata' => array('null' => true)]
+                                  )
+                                  ->getFirstItem();
+        
+        $productObject->setStoreId($storeId);
+        
+        return $productObject;
+    }
+    
+    private function checkForParentId($productId)
```
Comment (adamof) [inline Observer/Product.php]: This is too much for the observer, can you move this logic to a separate location?

File: Observer/Product.php [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+                    \Magento\Store\Model\ScopeInterface::SCOPE_STORE,
+                    $storeId
+                );
+        }
+        $storeIdConfigMap = array_unique($storeIdConfigMap);
+        
+        return array_keys($storeIdConfigMap);
+    }
+    
+    private function getProductWithRequestPath($product, $storeId) {
```
Comment (adamof) [inline Observer/Product.php]: with this as well

**Comment by nedick007** [comment]
@adamof ready for review

File: Controller/Adminhtml/Import/Ajax.php [hunk 3]
```diff
@@ -52,7 +53,11 @@ private function serializeRecords($records, $serializer) {
         $serializedData = [];
         
         foreach($records as $record) {
-            $serializedData[] = $serializer->serialize($record);
+            if($serializer->serialize($record)) {
+                $serializedData[] = $serializer->serialize($record);
```
Comment (adamof) [inline Controller/Adminhtml/Import/Ajax.php]: Why do you call `$serializer->serialize($record)` twice? Wouldn't that be unnecessarily expensive?

File: Helper/ProductSerializer.php [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        \Metrilo\Analytics\Model\ProductData                         $productData
+    ) {
+        $this->storeManager     = $storeManager;
+        $this->bundleType       = $bundleType;
+        $this->groupedType      = $groupedType;
+        $this->configurableType = $configurableType;
+        $this->productData      = $productData;
+    }
+    
+    public function serialize($product)
```
Comment (adamof) [inline Helper/ProductSerializer.php]: Would you enjoy reading this code? Look at it and tell me some potential problems with it.

File: Helper/ProductSerializer.php [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    }
+    
+    protected function getParentId($productId)
+    {
+        return $this->configurableType->getParentIdsByChild($productId) || $this->bundleType->getParentIdsByChild($productId) || $this->groupedType->getParentIdsByChild($productId);
+    }
+    
+    protected function checkForParentId($productId)
+    {
+        if ($this->configurableType->getParentIdsByChild($productId)) {
```
Comment (adamof) [inline Helper/ProductSerializer.php]: How is this different from the code on lines 64-70?

File: Observer/Product.php [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        \Metrilo\Analytics\Helper\ProductSerializer        $productSerializer,
+        \Magento\Framework\App\Config\ScopeConfigInterface $scopeConfig
+    ) {
+        $this->helper            = $helper;
+        $this->apiClient         = $apiClient;
+        $this->productSerializer = $productSerializer;
+        $this->scopeConfig       = $scopeConfig;
+    }
+    
+    private function getStoreIdsPerProject($storeIds) {
```
Comment (adamof) [inline Observer/Product.php]: isn't this duplicated in Category observer?!

**Comment by nedick007** [comment]
@adamof ready for review

File: Helper/ProductSerializer.php (hunk 4 above)
Comment (adamof) [reply Helper/ProductSerializer.php]: You haven't answered this? :)

File: Helper/ProductSerializer.php [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        $this->productData      = $productData;
+    }
+    
+    public function serialize($product)
+    {
+        $productId     = $product->getId();
+        $productType   = $product->getTypeId();
+        $storeId       = $product->getStoreId();
+    
+        if ($product->getData('observer_update')) {
```
Comment (adamof) [inline Helper/ProductSerializer.php]: I don't think it is any of the Serializer's business to make these decisions for parent/child products, can this be cleared from the serializer? Tell me what do you think

Comment (nedick007) [reply Helper/ProductSerializer.php]: I can move all this parent/child logic to additional productRelation helper for example and just return the productOptions data, if any.

**Comment by nedick007** [comment]
@adamof bump

File: Helper/ProductSerializer.php (hunk 7 above)
Comment (adamof) [reply Helper/ProductSerializer.php]: This is still unsolved

Comment (nedick007) [reply Helper/ProductSerializer.php]: All product options logic is now moved to separate helper classes for product options and product image respectively.

**Comment by nedick007** [comment]
@adamof bump

**Comment by nedick007** [comment]
@adamof I think i managed to resolve all the troublesome import scenarios like product catalog visibility and double parent products.

**Comment by nedick007** [comment]
@adamof bump

File: Helper/Data.php [hunk 8]
```diff
@@ -209,4 +212,22 @@ public function requestLogger($loggerPath, $loggerData) {
         file_put_contents($loggerPath, $loggerData, FILE_APPEND);
         file_put_contents($loggerPath, PHP_EOL, FILE_APPEND);
     }
+    
+    public function getStoreIdsPerProject($storeIds) {
```
Comment (adamof) [inline Helper/Data.php]: does this return store ids?

File: Helper/ProductOptions.php [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+                'imageUrl' => $imageUrl
+            ];
+        }
+        
+        return $productOptions;
+    }
+    
+    public function checkForParentId($productId)
+    {
+        return $this->configurableType->getParentIdsByChild($productId) || $this->bundleType->getParentIdsByChild($productId) || $this->groupedType->getParentIdsByChild($productId);
```
Comment (adamof) [inline Helper/ProductOptions.php]: Why do you check for the bundle or grouped type here?

File: Helper/Data.php (hunk 8 above)
Comment (nedick007) [reply Helper/Data.php]: Yes.

File: Helper/ProductOptions.php (hunk 9 above)
Comment (nedick007) [reply Helper/ProductOptions.php]: To exclude the products with parents from sync, but only configurable products with parents should be excluded, so this function is removed now.

**Comment by nedick007** [comment]
@adamof bump

**Pull request merged**